pennylane/optimize/rotoselect.py

```diff
@@ -0,0 +1,137 @@
+    optimized over the parameter :math:`\theta_d`. :math:`\text{arctan2}(x, y)` computes the
+    element-wise arc tangent of :math:`x/y` choosing the quadrant correctly, avoiding, in
+    particular, division-by-zero when :math:`y = 0`.
+
+    The algorithm is described in further detail in `Ostaszewski et al. (2019) <https://arxiv.org/abs/1905.09692>`_
+
+    Keyword Args:
+        possible_generators (list): List containing the possible ``pennylane.ops.qubit`` operators
+            that are allowed in the circuit. Default is the set of Pauli matrices :math:`\{X, Y, Z\}`.
+    """
+
+    def __init__(self, **kwargs):
+        self.possible_generators = kwargs.get("possible_generators", [qml.RX, qml.RY, qml.RZ])
+
+    def step(self, objective_fn, x, generators):
+        """Update x with one step of the optimizer.
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            x (array): NumPy array containing the current values of the variables to be updated
+            generators (list): List containing ``pennylane.ops.qubit`` operators to be used in the
+                circuit and updated
+
+        Returns:
+            array: the new variable values :math:`x^{(t+1)}` as well as the new generators.
+        """
+        # make sure that x is an array
+        if np.ndim(x) == 0:
+            x = np.array([x])
+
+        for d, _ in enumerate(x):
+            x[d], generators[d] = self._find_optimal_params(objective_fn, x, generators, d)
+
+        return x, generators
+
+    def update_possible_generators(self, possible_generators):
+        """Update list of possible generators.
+
+        Args:
+            possible_generators (list): list of new possible generators to be used in optimization
+        """
+        self.possible_generators = possible_generators
+
+    def _find_optimal_params(self, objective_fn, x, generators, d):
+        """Optimizer for the generators.
+
+        Optimizes for the best generator at position ``d``.
+
+        Args:
+            objective_fn (function): the objective function for optimization
+            x (array): NumPy array containing the current values of the variables to be updated
+            generators (list): List containing ``pennylane.ops.qubit`` operators to be used in the
+                circuit and to be updated
+            d (int): the position in the input array ``x`` containing the value to be optimized
+
+        Returns:
+            tuple: Tuple containing the parameter value and generator that, at position ``d`` in their
+                respective lists ``x`` and ``generators``, optimizes the objective function
+        """
+        for i, generator in enumerate(self.possible_generators):
+            generators[d] = generator
+
+            x = self._rotosolve(objective_fn, x, generators, d)
+            params_cost = objective_fn(x, generators)
```

Comment:
In the case of the rotoselect optimizer, does the objective function provided by the user have to have the signature `cost(x, generators)`? In this case, this is definitely something that should be mentioned in the docstring for `step`, as well as in the example in the class docstring

Reply:
Yes, that's correct. The cost function for rotoselect should be ``cost(x, generators)``. I'll add a comment about that to the docstring.

Reply:
Looks much better @thisac! I might make it even more explicit though, i.e., 'the objective function must have the signature `objective_fn(x, generators=None)` where...'